ably/realtime_conn.go

```diff
@@ -15,26 +16,40 @@ var (
 	errQueueing = errors.New("unable to send messages in current state with disabled queueing")
 )
 
+// ConnectionMode is the mode in which the connection is operating
+type ConnectionMode uint
+
+const (
+	// NormalMode this is set when the Connection operating normally
+	NormalMode ConnectionMode = iota
+	// ResumeMode this is set when the Connection is trying to resume
+	ResumeMode
+	// RecoveryMode this is set when the Connection is trying to recover
+	RecoveryMode
+)
+
 // Connection represents a single connection Realtime instantiates for
 // communication with Ably servers.
 type Connection struct {
-	details   proto.ConnectionDetails
-	id        string
-	serial    int64
-	msgSerial int64
-	err       error
-	conn      proto.Conn
-	msgCh     chan *proto.ProtocolMessage
-	opts      *ClientOptions
-	state     *stateEmitter
-	stateCh   chan State
-	pending   pendingEmitter
-	queue     *msgQueue
-	auth      *Auth
-
-	callbacks    connCallbacks
-	reconnecting bool
+	id          string
+	key         string
+	recoveryKey string
+	serial      int64
+	msgSerial   int64
+	err         error
+	conn        proto.Conn
+	msgCh       chan *proto.ProtocolMessage
+	opts        *ClientOptions
+	state       *stateEmitter
+	stateCh     chan State
+	pending     pendingEmitter
+	queue       *msgQueue
+	auth        *Auth
+
+	callbacks connCallbacks
+	mode      ConnectionMode
```

Comment:
I don't think this belongs here. What you call connection mode only affects connection establishment, and can be encapsulated there.

Reply:
addressed in 4449023b755bb04b26b63f4dab7c83ccd5dba667